### PR TITLE
osd: put peering events in main sharded wq

### PR DESCRIPTION
**Pull request by liewegas**
This didn't end up being too hard to get working.   There are two interesting things that happened:

1. The old peering wq would grab a pile of PGs, process them all, and then send notify/query/info messages with batched info. Putting this in the main wq means we process each event individually and we'll have a message per thing.  For large clusters this won't make any difference, since in general two osds won't share more than one pg, but for small clusters it will mean (1) more smaller messages and (2) slightly lower latency (since we send them immediately).

2. The old wq had a limit on the number of osdmaps that would be processed; once it hit the limit it would bail out and requeue the event.  Since the OSD-level queue was just a list of PGs with events queued it turned things into a rough-but-imprecise breadth-first osdmap consumption thing that improved osdmap cache locality.  We can't really do that here since the events are in the outer queue and ordered.  Instead, I just made  handle_osd_map block until pgs have consumed the new maps up tot the configured allowed-lagginess threshold using the existing pg epoch lower bound tracking.  

Thoughts/concerns?

Next steps after this (probably after merge) would be:

- make all of the messages fast dispatch, delivered directly into the sharded wq.  This is where the bigger win will be as (I think?) it will get rid of a bunch of the gross waiting_for_split machinery which is basically dupilcating the waiting_for_pg stuff in the sharded wq.
- more general cleanup. skimming the code I think there is a lot to be improved wrt memory allocations and such with the peering events (can the PGPeeringEvent be place directly into the PGpeeringItem instead of a shared_ptr to another thing), RecoveryCtx, etc.
- continue with the PG interface cleanup (the peering event was the thing I ran up against when i last looked at it)

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
Jenkins test this please

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -7720,38 +7720,24 @@ void OSD::check_osdmap_features()
...
   for (;
-       next_epoch <= osd_epoch && (no_max || next_epoch <= max);
+       next_epoch <= osd_epoch;
        ++next_epoch) {
     OSDMapRef nextmap = service.try_get_map(next_epoch);
     if (!nextmap) {
       dout(20) << __func__ << " missing map " << next_epoch << dendl;
-      // make sure max is bumped up so that we can get past any
-      // gap in maps
-      max = MAX(max, next_epoch + cct->_conf->osd_map_max_advance);
```
Comment (gregsfortytwo) [inline src/osd/OSD.cc]: Okay, it looks like this max comes from http://tracker.ceph.com/issues/7576, so it's entirely about the in-memory osdmap consumption and freshness of PGs.
So I guess maybe it's okay to just do every PG in one go?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -7277,6 +7277,17 @@ void OSD::handle_osd_map(MOSDMap *m)
...
   }
 
+  // wait for pgs to catch up
+  {
+    epoch_t num = last - first + 1;
+    if (cct->_conf->osd_map_cache_size > num &&
+	cct->_conf->osd_map_cache_size < last) {
+      epoch_t min = last - cct->_conf->osd_map_cache_size + 1;
+      dout(10) << __func__ << " waiting for pgs to consume " << min << dendl;
+      service.wait_min_pg_epoch(min);
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/osd/OSD.cc]: I don't think we can stall progress; the PG code never blocks on anything other than backend IO.

Generally speaking, we have two choices: (1) ingest maps without bound or throttling and let the PGs get arbitrarily behind, or (2) keep PGs caught up to the cache (but potentially have a queue of newer map updates, up to whatever the incoming message throttle size is).  I think (1) is better for performance (we ensure cache will work).  The behavior might be slightly more wonky for a laggy OSD (a tell command might be slow?) but I think the simplicity is worth it.

File: src/osd/OSD.cc (hunk 1 above)
Comment (liewegas) [reply src/osd/OSD.cc]: Right.  When I looked at this more carefully there are only a few cases that can cause a PG to get behind here:

- a wonkified advance_map on code prior to this patch.  in that case, the "do N maps per pass through advance_map" could get out of sync for different PGs so that they weren't all working on the same range of epochs (e.g., because we were doing that, then crashed, getting the ranges out of sync).  this PR erases that possibility, so that the only time we would hit it woudl be the very first startup (before we are marked up) after an upgrade.
- an imported PG.  ceph-objectstore-tool now catches it up to the latest osdmap epoch, so this isn't possible any more.

File: src/common/legacy_config_opts.h [hunk 3]
```diff
@@ -874,6 +874,7 @@ OPTION(mon_rocksdb_options, OPT_STR)
  */
 OPTION(osd_client_op_priority, OPT_U32)
 OPTION(osd_recovery_op_priority, OPT_U32)
+OPTION(osd_peering_op_priority, OPT_U32)
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/common/legacy_config_opts.h]: lazyness and procrastination. it will be easier to clean these all up at once.

[the remark this answers is not in the document]
Comment (liewegas) [reply src/common/legacy_config_opts.h]: (and we can't just call get_val<> instead because it is slower (mutex + map<> lookup))

File: src/osd/OSD.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/osd/OSD.cc]: Did you mean you think (2) is better for performance? Otherwise that doesn't make sense.

And yes, I think (2) is the way to go here. But I'm worried because even if it's not an actual deadlock, it means a slow PG can prevent active PGs from doing stuff like, well, peering. Which seems like a serious issue!

[the remark this answers is not in the document]
Comment (liewegas) [reply src/osd/OSD.cc]: Yeah I meant 2, sorry :)

It means that a slowly peering PG can prevent injest of new maps... which could prevent peering (and various other stuff, yeah).  This doesn't concern me in the general sense (the work that advmap does is very lightweight) *but* if something is stuck somewhere holding a PG lock (e.g., hung read, deadlock, bug) then everything else will eventually grind to a halt too because the map consumption is blocked.  Until we suicide due to the failed heartbeat on the stuck PG thread.

In that case, though, it seems like the right mitigation is to perhaps trigger a suicide earlier if it is holding up map consumption? Seems like a secondary concern, though ("we could have detected our internal deadlock sooner").  And easy to fix with a similar (perhaps shorter?) heartbeat on the main dispatch thread.

**Comment by liewegas** [comment]
Ok, this has been updated with a new message throttling and cache pinning strategy:

1. We pin osdmaps in the cache based on the pg epoch lower bound, so that they are never starved by the cache.  
2. There is a new config option that controsl what multiple of the map cache we allow the pgs to lag behind before we throttle ingest of new maps.  Set to 2x by default; happy to use whatever value we want here.

**Comment by gregsfortytwo** [comment]
@liewegas, still not passing make check :(

File: src/osd/PG.cc [hunk 4]
```diff
@@ -7381,7 +7382,8 @@ boost::statechart::result PG::RecoveryState::Active::react(const MInfoRec& infoe
   // don't update history (yet) if we are active and primary; the replica
   // may be telling us they have activated (and committed) but we can't
   // share that until _everyone_ does the same.
-  if (pg->is_actingbackfill(infoevt.from)) {
+  if (pg->is_actingbackfill(infoevt.from) &&
+      pg->peer_activated.count(infoevt.from) == 0) {
```
Comment (gregsfortytwo) [inline src/osd/PG.cc]: Is this from "osd/PG: ignore dup activation messages" an unrelated bug fix?

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -7347,6 +7349,25 @@ void OSD::handle_osd_map(MOSDMap *m)
...
+  {
+    // we extend the map cache pins to accomodate pgs slow to consume maps
+    // for some period, until we hit the max_lag_factor bound, at which point
+    // we block here to stop injesting more maps than they are able to keep
+    // up with.
+    epoch_t max_lag = cct->_conf->osd_map_cache_size *
+      m_osd_pg_epoch_max_lag_factor;
+    assert(max_lag > 0);
+    if (osdmap->get_epoch() > max_lag) {
+      epoch_t min = osdmap->get_epoch() - max_lag;
```
Comment (gregsfortytwo) [inline src/osd/OSD.cc]: So we're entering this block whenever the osdmap epoch is bigger than our max lag (so, always, after we have run for a little while). Are we just counting on the service.wait_min_pg_epoch() call to return instantly in those cases?
It does, other than grabbing the lock, so it works. The code structure and logging output led me to think we'd only do the call if we thought we'd exceeded our bounds, though; we shouldn't output at level 10 for a non-"failure" for instance...

File: src/common/options.cc [hunk 6]
```diff
@@ -2617,6 +2617,10 @@ std::vector<Option> get_global_options() {
     .set_default(0)
     .set_description(""),
 
+    Option("osd_pg_epoch_max_lag_factor", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
```
Comment (gregsfortytwo) [inline src/common/options.cc]: Shouldn't this get placed with the osd_map_cache_size, since it's a modifier on that?
And it should use the .add_see_also() stuff?

File: src/osd/OSD.cc (hunk 5 above)
Comment (liewegas) [reply src/osd/OSD.cc]: fixed this message to be more informative

File: src/osd/PG.cc (hunk 4 above)
Comment (liewegas) [reply src/osd/PG.cc]: yeah, unrelated fix for issue i encountered while testing this

File: src/common/options.cc (hunk 6 above)
Comment (liewegas) [reply src/common/options.cc]: fixed

**Comment by liewegas** [comment]
@gregsfortytwo this is passing my tests now!   i think it's ready to go
